pipelineRunner.go

```diff
@@ -50,712 +49,527 @@ func NewPipelineRunner(envvarHelper EnvvarHelper, whenEvaluator WhenEvaluator, d
+
+	pr.sendStatusMessage(stage, parentStageName, depth, runIndex, nil, &runDuration, finalStatus)
+
+	if len(stage.Services) > 0 {
+		// this stage has service containers, stop them now that the stage has finished
+		pr.dockerRunner.StopServiceContainers(ctx, stage)
 	}
 
-	pr.tailLogsChannel <- contracts.TailLogLine{
-		Step:        p.Name,
-		ParentStage: parentStageName,
-		Type:        "stage",
-		Depth:       depth,
-		RunIndex:    runIndex,
-		Duration:    &result.DockerRunDuration,
-		ExitCode:    &result.ExitCode,
-		Status:      &status,
+	return
+}
+
+func (pr *pipelineRunnerImpl) RunStageWithRetry(ctx context.Context, depth int, dir string, envvars map[string]string, parentStage *manifest.EstafetteStage, stage manifest.EstafetteStage) (err error) {
+
+	span, ctx := opentracing.StartSpanFromContext(ctx, "RunStageWithRetry")
+	defer span.Finish()
+	span.SetTag("stage", stage.Name)
+
+	runIndex := 0
+	retries := stage.Retries
+	if (len(stage.ParallelStages) > 0 || len(stage.Services) > 0) && retries > 0 {
+		// retries are not supported in combination with parallel stages or services for now
+		retries = 0
 	}
 
-	if result.DockerRunError != nil {
-		err = result.DockerRunError
+	parentStageName := ""
+	if parentStage != nil {
+		parentStageName = parentStage.Name
 	}
 
-	if result.Canceled {
-		log.Info().Msgf("[%v] Canceled pipeline '%v'", p.Name, p.Name)
-	} else if err != nil {
-		log.Warn().Err(err).Msgf("[%v] Pipeline '%v' container failed", p.Name, p.Name)
-	} else {
-		log.Info().Msgf("[%v] Finished pipeline '%v' successfully", p.Name, p.Name)
+	// retry until successful or number of retries is maxed out
+	for runIndex <= retries {
+		err = pr.RunStage(ctx, depth, runIndex, dir, envvars, parentStage, stage)
+
+		// if canceled during stage stop further execution
+		if pr.canceled {
+			return nil
+		}
+
+		// if execution is successful, we're done
+		if err == nil {
+			return nil
+		}
+
+		// create log line for error
+		logLineObject := contracts.BuildLogLine{
+			Timestamp:  time.Now().UTC(),
+			StreamType: "stderr",
+			Text:       err.Error(),
+		}
+
+		pr.tailLogsChannel <- contracts.TailLogLine{
+			Step:        stage.Name,
+			ParentStage: parentStageName,
+			Type:        contracts.TypeStage,
+			Depth:       depth,
+			RunIndex:    runIndex,
+			LogLine:     &logLineObject,
+		}
+
+		runIndex++
 	}
 
-	if len(p.Services) > 0 {
-		// this stage has service containers, stop them now that the stage has finished
-		go pr.dockerRunner.stopServices(ctx, &p, p.Services)
+	if err != nil {
+		return err
 	}
 
-	return
+	return nil
 }
 
-func (pr *pipelineRunnerImpl) runService(ctx context.Context, envvars map[string]string, parentStage *manifest.EstafetteStage, service manifest.EstafetteService) (result estafetteServiceRunResult, err error) {
+func (pr *pipelineRunnerImpl) RunService(ctx context.Context, envvars map[string]string, parentStage manifest.EstafetteStage, service manifest.EstafetteService) (err error) {
```

Comment:
it looks like the code here is really similar to RunStage.

It's outside of the scope of this PR, but do you think reusing some of the code would help in consistency? Or is it adding more complexity?

Reply:
It's pretty similar but different on key parts; I guess the similar parts can be extracted in other functions and be reused.

Reply:
Extracted similar parts into shared functions. A lot is passed as pointers, so that isn't particularly nice but it saves a lot of duplication.